Update the NoSQL image and documentation (#2398)
oracle/docker-images: NoSQL/README.md

* [Oracle NoSQL Database Community Edition](ce/Dockerfile)

> This container image uses KVLite. **KVLite is a simplified version of the Oracle NoSQL Database**. It provides a single storage node, a single shard store, with no
> replication and no administration. It runs in a single process. **Note:** KVLite is not intended for production deployment or performance measurements.

Reviewer: ```suggestion
This container image uses a simplified version of the Oracle NoSQL Database called KVLite. KVLite runs as a single process that provides a single storage node and single storage shard. KVLite does not include replication or administration. 

> **Note:** KVLite is not intended for production deployment or performance measurements.
```

```shell
docker run -d --name=kvlite --hostname=kvlite --env KV_PROXY_PORT=8080 -p 8080:8080 oracle/nosql:ce
```
**Note**:  By default, the KVLite store created has a size of `10GB`, use `--env KV_STORAGESIZE=N` to override that. 

> The value of N is in GB and must be >= 1

Reviewer: ```suggestion
The value of `N` is in gigabytes and must be greater than 1GB.
```

});
````

## Advanced Scenario: Using Oracle NoSQL Command-Line from an external host

> **Note**: We recommend running NoSQL Command-Line doing a container to container connection as shown in the previous chapters. 
> It allows starting the container without publishing all internal ports (KVPORT, KV_HARANGE, KV_SERVICERANGE) but only the KV_PROXY_PORT. 

Reviewer: ```suggestion
**Note**: We recommend running NoSQL Command-Line and creating a container-to-container connection as detailed above.. 

It allows you to start a container with only the `KV_PROXY_PORT` enabled. 
```

For your developments, remember the SDK drivers will contact the Oracle NoSQL Database Proxy on KV_PROXY_PORT. 

If you need to run NoSQL Command-Line from a host outside any container, please follow those instructions.

Install Oracle NoSQL in your external host

```shell
KV_VERSION=21.2.46
rm -rf kv-$KV_VERSION
DOWNLOAD_ROOT=http://download.oracle.com/otn-pub/otn_software/nosql-database
DOWNLOAD_FILE="kv-ce-${KV_VERSION}.zip"
DOWNLOAD_LINK="${DOWNLOAD_ROOT}/${DOWNLOAD_FILE}"
curl -OLs $DOWNLOAD_LINK
jar tf $DOWNLOAD_FILE | grep "kv-$KV_VERSION/lib" > extract.libs
jar xf $DOWNLOAD_FILE @extract.libs 
rm -f $DOWNLOAD_FILE extract.libs
KVHOME=$PWD/kv-$KV_VERSION
```

Start up KVLite in a container. You must give it a name and provide a hostname. 

> In this case, You need to publish all internal ports and the KV_PROXY_PORT.

Reviewer: ```suggestion
In this case, You need to publish all internal ports and the `KV_PROXY_PORT`:
```

> - 5000 KVPORT
> - 5010-5020 KV_HARANGE
> - 5021-5049 KV_SERVICERANGE
> - 8080 KV_PROXY_PORT

Reviewer: ```suggestion

- 5000 `KVPORT`
- 5010-5020 `KV_HA_RANGE` 
- 5021-5049 `KV_SERVICE_RANGE`
- 8080 `KV_PROXY_PORT`
```

Startup of KVLite is the default `CMD` of the image:

> You will use you current HOSTNAME as a value for the --hostname

Reviewer: ```suggestion
Use the environment variable `$HOSTNAME` as the value for the `--hostname`
```

```shell
docker run -d --name=kvlite --hostname=$HOSTNAME --env KV_PROXY_PORT=8080 -p 8080:8080 \
-p 5000:5000 -p 5010-5020:5010-5020 -p 5021-5049:5021-5049 -p 5999:5999 oracle/nosql:ce
```

**Note**:  By default, the KVLite store created has a size of `10GB`, use `--env KV_STORAGESIZE=N` to override that. 

> The value of N is in GB and must be >= 1

Reviewer: ```suggestion
The value of `N` is in gigabytes and must be greater than 1GB.
```

In a second shell, run the NoSQL command to ping the kvlite store
instance:

```shell
java -jar $KVHOME/lib/kvstore.jar ping -host $HOSTNAME -port 5000
```

> Note: -host must be the same name used when starting the container

Reviewer: ```suggestion
Note: the value provided for `-host` must match the name used when starting the container.
```

> If you want to run the NoSQL command to ping the kvlite store from another container:

Reviewer: ```suggestion
If you want to run the NoSQL command to ping the KVLite store from another container:
```

```shell
docker run --rm -ti --link kvlite:store oracle/nosql:ce \
  java -jar lib/kvstore.jar ping -host store -port 5000
```

> Note the required use of --link for proper hostname check (actual KVLite container is named kvlite; alias is store).

Reviewer: ```suggestion
Note the required use of `--link` for proper hostname check (actual KVLite container is named `kvlite` and its alias is `store`).
```

> If you want to run without --link, you cannot use any alias when starting the container (use HOSTNAME).  

Reviewer: ```suggestion
If you want to run without `--link`, you can not use any alias when starting the container, so use `$HOSTNAME` instead..  
```

64 bytes from nosql-container-host (10.0.0.143): icmp_seq=3 ttl=64 time=0.192 ms
```

```shell
docker run -d --name=kvlite --hostname=kvlite-nosql-container-host --env KV_PROXY_PORT=8080 -p 8080:8080 \
-p 5000:5000 -p 5010-5020:5010-5020 -p 5021-5049:5021-5049 -p 5999:5999 oracle/nosql:ce
```
**Note**:  By default, the KVLite store created has a size of `10GB`, use `--env KV_STORAGESIZE=N` to override that. 

> The value of N is in GB and must be >= 1

Reviewer: ```suggestion
The value of `N` is in gigabytes and must be greater than 1GB.
```

```shell
java -jar $KVHOME/lib/kvstore.jar ping -host kvlite-nosql-container-host -port 5000
```

> Note: -host must be the same name used when starting the container

Reviewer: ```suggestion
Note: `-host` must be the same name used when starting the container
```

You can also use the admin Oracle NoSQL Command Line Interface (CLI).

```shell
java -jar $KVHOME/lib/kvstore.jar runadmin -host kvlite-nosql-container-host -port 5000 -store kvstore
````

You can also use the Oracle SQL Shell Command Line Interface (CLI)

```shell
java -jar $KVHOME/lib/sql.jar -helper-hosts kvlite-nosql-container-host:5000 -store kvstore
````

## Quick start: building the Oracle NoSQL Community Edition image

To build the Oracle NoSQL Community Edition container image, clone this
repository and run the following commands from the root of cloned repository:

```shell
cd NoSQL/ce/
docker build -t oracle/nosql:ce .
```
or

```shell
cd NoSQL/ce/
docker build --build-arg KV_VERSION=21.2.46 --tag oracle/nosql:ce .
```

The resulting image will be available as `oracle/nosql:ce`. 

## More information

For more information on [Oracle NoSQL](http://www.oracle.com/technetwork/database/database-technologies/nosqldb/overview/index.html)

> please review the [product documentation](https://docs.oracle.com/en/database/other-databases/nosql-database/index.html).

Reviewer: ```suggestion
please review the [Oracle NoSQL Database product documentation](https://docs.oracle.com/en/database/other-databases/nosql-database/index.html).
```

The Oracle NoSQL Database Community Edition image contains the OpenJDK.

## Licenses

> Oracle NoSQL Community Edition is licensed under the [APACHE LICENSE v2.0](https://docs.oracle.com/en/database/other-databases/nosql-database/22.1/license/index.html#NSXLI-GUID-006E432E-1965-45A2-AEDE-204BD05E1560) . 

Reviewer: ```suggestion
Oracle NoSQL Community Edition is released under the [Apache 2.0 License](https://docs.oracle.com/en/database/other-databases/nosql-database/22.1/license/index.html#NSXLI-GUID-006E432E-1965-45A2-AEDE-204BD05E1560) . 
```
